### PR TITLE
Health API fix the reporting of initializing shards

### LINKED ISSUE #90327: [HealthAPI] Shards availability indicator doesn't report diagnosis when disk is out of space
### Elasticsearch Version

8.4, 8.5, 8.6

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

The shards availability indicator should always be reporting diagnosis when its status is RED/YELLOW.
However, when an index is on a data node that runs out of disk space it will report RED but no diagnosis.

### Steps to Reproduce

Have multi-node cluster where a data (non-master) node that hosts an index runs out of disk space (the index cannot relocate somewhere else).
Call the Health API and the output will be something along these lines:
```
    "shards_availability": {
      "status": "red",
      "symptom": "This cluster has all shards available.",
      "details": {
        "initializing_primaries": 1,
        "started_replicas": 0,
        "initializing_replicas": 0,
        "unassigned_primaries": 0,
        "started_primaries": 1,
        "creating_primaries": 0,
        "unassigned_replicas": 0,
        "restarting_primaries": 0,
        "restarting_replicas": 0
      },
      "impacts": [
        {
          "id": "elasticsearch:health:shards_availability:impact:primary_unassigned",
          "severity": 1,
          "description": "Cannot add data to 1 index [test_index]. Searches might return incomplete results.",
          "impact_areas": [
            "ingest",
            "search"
          ]
        }
      ] 
    }
``` 

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by gmarouli**
Fixes: https://github.com/elastic/elasticsearch/issues/90327

### Unexpected Behaviour
During testing we encountered the following unexpected health API response:
```
    "shards_availability": {
      "status": "red",
      "symptom": "This cluster has all shards available.",
      "details": {
        "initializing_primaries": 1,
        "started_replicas": 0,
        "initializing_replicas": 0,
        "unassigned_primaries": 0,
        "started_primaries": 1,
        "creating_primaries": 0,
        "unassigned_replicas": 0,
        "restarting_primaries": 0,
        "restarting_replicas": 0
      },
      "impacts": [
        {
          "id": "elasticsearch:health:shards_availability:impact:primary_unassigned",
          "severity": 1,
          "description": "Cannot add data to 1 index [test_index]. Searches might return incomplete results.",
          "impact_areas": [
            "ingest",
            "search"
          ]
        }
      ] 
    }
```
This is an unexpected response because it has inconsistencies:
- The status is `RED`
- The impact mentions about an unassigned primary shard
However,
- The details report a started shard and one initializing
- The symptom mentions that everything is fine.

### Reproduction Path
We were not able to reproduce it locally but we were able to track down from where could such an inconsistency occur. 

**Cause**
We believe that the cause is on the way the above attributes are calculated. Both the status and the impact use the following condition (routing.active() || isRestarting || isNew) == false. From the following code snippet:
```
            boolean isNew = isUnassignedDueToNewInitialization(routing);
            boolean isRestarting = isUnassignedDueToTimelyRestart(routing, shutdowns);
            available &= routing.active() || isRestarting || isNew;
            if ((routing.active() || isRestarting || isNew) == false) {
                indicesWithUnavailableShards.add(routing.getIndexName());
            }
```
I believe this is where the problem is, it doesn't account for an initializing shard. The shard is not active and it's not unassigned so the condition results to false. The symptom, diagnoses and details are looking for unassigned shards which we currently do not have. 

### Proposal 1
~Assuming having a shard that is INITIALIZING is part of the happy flow and the user cannot do anything to fix this than wait, I propose to take this into consideration and to not consider it as a health problem.~

### Proposal 2
Considering that an `INITIALIZING` shard has impact for the user, the health API should report that. We propose the following:
```
    "shards_availability": {
      "status": "red",
      "symptom": "This cluster has 1 initializing shard.",
      "details": {
        "initializing_primaries": 1,
        "started_replicas": 0,
        "initializing_replicas": 0,
        "unassigned_primaries": 0,
        "started_primaries": 1,
        "creating_primaries": 0,
        "unassigned_replicas": 0,
        "restarting_primaries": 0,
        "restarting_replicas": 0
      },
      "impacts": [
        {
          "id": "elasticsearch:health:shards_availability:impact:primary_unassigned",
          "severity": 1,
          "description": "Cannot add data to 1 index [test_index]. Searches might return incomplete results.",
          "impact_areas": [
            "ingest",
            "search"
          ]
        }
      ] 
      "diagnoses": [
        {
          "id": "elasticsearch:health:shards_availability:diagnosis:wait_for_initialization",
          "action": "Please wait until the initialization of the shards is finished.",
          "cause": "Elasticsearch is currently initializing one or more shards.",
          "help_url": <a new guide that helps them follow the progress maybe?>
          "affected_resources": ....
        }
      ] 
    }
```

Notes:
_Should we rename the `elasticsearch:health:shards_availability:impact:primary_unassigned` to `elasticsearch:health:shards_availability:impact:primary_unavailable`?_
We decided against this, to avoid unforeseen bugs if this is references somewhere on the cloud. This is an id, the message contains all the information that are useful for humans, and there the message is correct.

_Should we introduce a global `wait` diagnosis? Something like "Elasticsearch is recovering, please give it time"_
Yes we already have something like this, so it is not new.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java [hunk 1]
```diff
@@ -556,10 +556,10 @@ List<Diagnosis.Definition> diagnoseAllocationResults(
      * @return A predicate that returns true if the decision exists and matches the expected outcome, false otherwise.
      */
     private static Predicate<NodeAllocationResult> hasDeciderResult(String deciderName, Decision.Type outcome) {
-        return (nodeResult) -> nodeResult.getCanAllocateDecision()
-            .getDecisions()
-            .stream()
-            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+        return (nodeResult) -> {
+            Decision decision = nodeResult.getCanAllocateDecision();
+            return decision != null && decision.getDecisions().stream().anyMatch(d -> deciderName.equals(d.label()) && outcome == d.type());
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:561]: This was a bug I encountered during my efforts to reproduce this issue. The decision can be `null`.

**Comment by gmarouli** [comment]
Hi @idegtiarenko , I added you as a reviewer since you have worked a lot on this. Please take a look if you have time, of course.

**Comment by gmarouli** [comment]
After an offline discussion with @idegtiarenko , we see that handling an initializing shards is not as simple as what this PR was initially implying. A shard is initializing during a recovery of some sort. During this time there is impact for the user since that shard is not available, either the primary shard is not available meaning the searches will be incomplete or a replica shard meaning the searches might perform worse.

Both of the above should not be considered `green`, in order to be inline with the way this indicator works it should be `red` and `yellow` respectively. However, the only action we have for the user is to wait for the shard to be initialized. 

I will look at the code and come up with a proposal for the diagnosis and impacts.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java [hunk 2]
```diff
@@ -392,8 +392,8 @@ private class ShardAllocationCounts {
         public void increment(ShardRouting routing, ClusterState state, NodesShutdownMetadata shutdowns, boolean verbose) {
             boolean isNew = isUnassignedDueToNewInitialization(routing);
             boolean isRestarting = isUnassignedDueToTimelyRestart(routing, shutdowns);
-            available &= routing.active() || isRestarting || isNew;
-            if ((routing.active() || isRestarting || isNew) == false) {
+            available &= routing.active() || isRestarting || isNew || routing.initializing();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:395]: I think the original logic is correct here. An initializing primary shard should result in RED health. Under any normal circumstance initializing primaries should not be expected.

The problem in my mind is with the way "symptom" is calculated, it should (probably) also show initializing shards. I am not 100% up to date on what symptom mean compared to status though.

Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java]: Agreed, we reverted this back to what it was and we fixed the symptom and introduced a new diagnosis.

**Review by idegtiarenko — APPROVED** [review]

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Comment by gmarouli** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java [hunk 3]
```diff
@@ -451,22 +468,16 @@ List<Diagnosis.Definition> diagnoseUnassignedShardRouting(ShardRouting shardRout
         List<Diagnosis.Definition> diagnosisDefs = new ArrayList<>();
         LOGGER.trace("Diagnosing unassigned shard [{}] due to reason [{}]", shardRouting.shardId(), shardRouting.unassignedInfo());
         switch (shardRouting.unassignedInfo().getLastAllocationStatus()) {
-            case NO_VALID_SHARD_COPY:
-                diagnosisDefs.add(ACTION_RESTORE_FROM_SNAPSHOT);
-                break;
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:456]: ah switch expressions ❤️ no more `break`s :)

**Review by andreidan — APPROVED** [review]
LGTM, thanks for fixing this

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java (hunk 3 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:456]: I couldn't resist 😇

**Comment by gmarouli** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.6|[](https://github.com/elastic/elasticsearch/pull/93586)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by craigtaverner** [comment]
Removed `backport pending` label due to https://github.com/elastic/elasticsearch/pull/93586

**Pull request merged**